internal/kubernetes/discovery.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2024 Crunchy Data Solutions, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package kubernetes
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/version"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/rest"
+
+	"github.com/crunchydata/postgres-operator/internal/logging"
+)
+
+type Version = version.Info
+
+// DiscoveryRunner implements [APIs] by reading from a Kubernetes API client.
+// Its methods are safe to call concurrently.
+type DiscoveryRunner struct {
+	Client interface {
+		ServerGroups() (*metav1.APIGroupList, error)
+		ServerResourcesForGroupVersion(string) (*metav1.APIResourceList, error)
+		ServerVersion() (*version.Info, error)
+	}
+
+	refresh time.Duration
+
+	want []API
```

Comment:
🤔 I can't come up with a really better term for this, but `want` had me thinking we want these object types to be present, when really, these are the GV/GK/GVK that we want details on if present.

Reply:
I'll see how `relevant` goes.